src/Microsoft.TemplateEngine.Edge/Installers/NuGet/NugetApiPackageManager.cs

```diff
@@ -173,54 +173,47 @@ private async Task<(PackageSource, IPackageSearchMetadata)> GetLatestVersionInte
+                await Task.WhenAll(
+                    packageSources.Select(source => GetPackageMetadataAsync(source, packageIdentifier, includePrerelease: true, cancellationToken)))
+                          .ConfigureAwait(false);
 
-            ConcurrentDictionary<PackageSource, IEnumerable<IPackageSearchMetadata>> searchResults = new ConcurrentDictionary<PackageSource, IEnumerable<IPackageSearchMetadata>>();
-            await Task.WhenAll(packageSources.Select(async source =>
-                {
-                    IEnumerable<IPackageSearchMetadata> foundPackages = await GetPackageMetadataAsync(source, packageIdentifier, includePrerelease: true, cancellationToken).ConfigureAwait(false);
-                    if (foundPackages == null)
-                    {
-                        return;
-                    }
-                    searchResults[source] = foundPackages;
-                })).ConfigureAwait(false);
-
-            if (!searchResults.Any())
+            if (!foundPackagesBySource.Any())
             {
                 throw new InvalidNuGetSourceException("Failed to load NuGet sources", packageSources.Select(source => source.Source));
             }
 
-            var accumulativeSearchResults = searchResults
-                .SelectMany(result => result.Value.Select(p => new { source = result.Key, package = p }));
+            var accumulativeSearchResults = foundPackagesBySource
+                .SelectMany(result => result.foundPackages.Select(package => (result.source, package)));
 
             if (!accumulativeSearchResults.Any())
             {
                 _nugetLogger.LogWarning($"{packageIdentifier} is not found in NuGet feeds {string.Join(", ", packageSources.Select(source => source.Source))}.");
                 throw new PackageNotFoundException(packageIdentifier, packageSources.Select(source => source.Source));
             }
 
-            var latestVersion = accumulativeSearchResults.Aggregate(
+            (PackageSource, IPackageSearchMetadata) latestVersion = accumulativeSearchResults.Aggregate(
                 (max, current) =>
                 {
-                    if (max == null) return current;
+                    if (max == default) return current;
                     return current.package.Identity.Version > max.package.Identity.Version ? current : max;
                 });
 
-            var latestStableVersion = accumulativeSearchResults.Aggregate(
+            (PackageSource, IPackageSearchMetadata) latestStableVersion = accumulativeSearchResults.Aggregate(
                 (max, current) =>
                 {
                     if (current.package.Identity.Version.IsPrerelease) return max;
-                    if (max == null) return current;
+                    if (max == default) return current;
                     return current.package.Identity.Version > max.package.Identity.Version ? current : max;
                 });
```

Comment:
These two queries can be merged into one, if you use `includePreview` inside the anonymous method. Then you can also get rid of the `if` check below.

Reply:
I'm not sure if this is possible - we need to check first if  the `latestStableVersion` is available. 
We can shuffle the statements around though to check for preview versions only if stable version is not available / not needed, but I'm not convinced for now if it makes code more readable - will try it out.

Reply:
Ah, you are right. My bad.